### PR TITLE
Begin Refactoring/redesigning

### PR DESCRIPTION
**Pull request by emerali**
TODO:

- [x] remove AbstractCoordinateLattice in favour of HasCoordinate trait

### CODE REVIEW

File: src/hypercubic.jl [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    clipped = []
+    for i in 1:N
+        c = _apply_bc(bcs[i], dims[i], site[i])
+        if isnothing(c)
+            return nothing
+        else
+            push!(clipped, c)
+        end
+    end
+    return SVector{N}(clipped)
```
Comment (Roger-luo) [inline src/hypercubic.jl:40]: just use vector, return `clipped` directly

File: src/hypercubic.jl [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        if sm < 1
+            return [sp]
+        else
+            return [sm, sp]
+        end
+    end
+end
+
+
+basis_vectors(lattice::HyperCubic{N}) where N = eachrow(SMatrix{N, N}(I))
```
Comment (Roger-luo) [inline src/hypercubic.jl:84]: just use `Diagonal`

**Review by Roger-luo — COMMENTED** [review]
I guess we can just use our own `Coordinate` type now.

File: src/boundaries.jl [hunk 3]
```diff
@@ -0,0 +1,37 @@
+abstract type AbstractBoundary end
+abstract type PrimitiveBoundary <: AbstractBoundary end
```
Comment (Roger-luo) [inline src/boundaries.jl:2]: This is not necessary?

File: src/hypercubic.jl [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+HyperCubic{N}(dims::NTuple{N, Int}, bc::B=Periodic()) where {N, B <: AbstractBoundary} = HyperCubic{N, B}(dims, bc)
+HyperCubic(dims::NTuple{N, Int}, bc::B=Periodic()) where {N, B <: AbstractBoundary} = HyperCubic{N, B}(dims, bc)
+
+HyperCubic{N, B}(dim::Int) where {N, B <: AbstractBoundary} = HyperCubic{N, B}(dim, B())
+HyperCubic{N}(dim::Int, bc::B=Periodic()) where {N, B <: AbstractBoundary} = HyperCubic{N, B}(dim, bc)
+
+const Chain = HyperCubic{1}
+const Square = HyperCubic{2}
+const SimpleCubic = HyperCubic{3}
```
Comment (Roger-luo) [inline src/hypercubic.jl:26]: Cubic

File: Project.toml [hunk 5]
```diff
@@ -4,6 +4,8 @@ authors = ["Roger-luo"]
 version = "0.2.2"
 
 [deps]
+DataFrames = "a93c6f00-e57d-5684-b7b6-d8193f3e46c0"
```
Comment (Roger-luo) [inline Project.toml:7]: rm this.

File: src/boundaries.jl (hunk 3 above)
Comment (Roger-luo) [reply src/boundaries.jl:2]:
```jl
const PrimitiveBoundaries = Union{Periodict, Open, Helical}
```

File: src/hypercubic.jl [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        end
+        new{N, B}(dims, bc)
+    end
+
+    function HyperCubic{N, B}(dim::Int, bc::B) where {N, B <: AbstractBoundary}
+        new{N, B}(ntuple(x->dim, N), bc)
+    end
+end
+
+HyperCubic{N, B}(dims::NTuple{N, Int}) where {N, B <: PrimitiveBoundary} = HyperCubic{N, B}(dims, B())
```
Comment (Roger-luo) [inline src/hypercubic.jl:16]:
```suggestion
HyperCubic{N, B}(dims::NTuple{N, Int}) where {N, B <: AbstractBoundary} = HyperCubic{N, B}(dims, B())
HyperCubic{N, MixedBoundary}(dims::NTuple{N, Int}) where {N} = error("")
```

File: src/boundaries.jl [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+# Helical BCs is a fast approximation to
+# Periodic BCs that takes advantage of the linear storage of
+# arrays. Those familiar with DMRG may recognize this as
+# the "snake" boundary condition.
+struct Helical <: PrimitiveBoundary end
+
+# Allows defining lattices with different BCs along each
+#  dimension. Should reject Helical BC as that is a "global" BC.
+struct MixedBoundary{N} <: AbstractBoundary
+    boundaries::NTuple{N, AbstractBoundary}
```
Comment (Roger-luo) [inline src/boundaries.jl:14]:
```suggestion
    boundaries::Vector{AbstractBoundary}
```

or 

```julia
struct MixedBoundary{N, BT <: Tuple}
    boundaries::BT
end
```

File: src/honeycomb.jl [hunk 8]
```diff
@@ -0,0 +1,7 @@
+
+struct HoneyComb <: AbstractLattice
+    dims::NTuple{2, Int}
+    boundaries::NTuple{2, AbstractBoundary}
```
Comment (Roger-luo) [inline src/honeycomb.jl:4]:
```suggestion
    boundaries::MixedBoundary{2}
```

File: src/hypercubic.jl [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        if bc isa MixedBoundary && !(bc isa MixedBoundary{N})
+            throw(ArgumentError("Given MixedBoundary should have same number of dimensions as the Lattice!"))
+        end
+        new{N, B}(dims, bc)
+    end
+
+    function HyperCubic{N, B}(dim::Int, bc::B) where {N, B <: AbstractBoundary}
+        new{N, B}(ntuple(x->dim, N), bc)
+    end
+end
```
Comment (Roger-luo) [inline src/hypercubic.jl:14]:
```suggestion
struct HyperCubic{N, B<:Tuple} <: AbstractLattice
    dims::NTuple{N, Int}
    bc::B
    function HyperCubic{N, B}(dims::NTuple{N, Int}, bc::NTuple{N, B}) where {N, B <: AbstractBoundary}
        new{N, B}(dims, bc)
    end

    function HyperCubic{N, B}(dim::Int, bc::B) where {N, B <: AbstractBoundary}
        new{N, B}(ntuple(x->dim, N), bc)
    end
end
```

File: src/hypercubic.jl [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+function _apply_bcs(lattice::HyperCubic{N, Periodic}, site::Coordinate{N, Int})::Coordinate{N, Int} where N
+    return Coordinate(_apply_bc.(Periodic, lattice.dims, site.coordinates)...)
+end
+
+function _apply_bcs(lattice::HyperCubic{N, MixedBoundary}, site::Coordinate{N, Int})::Union{Coordinate{N, Int}, Nothing} where N
+    dims = lattice.dims
+    bcs = lattice.bc.boundaries
+    coords = site.coordinates
+    clipped = []
+    for i in 1:N
```
Comment (Roger-luo) [inline src/hypercubic.jl:44]:
```suggestion
# NOTE: use @nexprs etc. to expand this.
        for i in 1:N
```

File: src/hypercubic.jl [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+HyperCubic(dims::NTuple{N, Int}, bc::B=Periodic()) where {N, B <: AbstractBoundary} = HyperCubic{N, B}(dims, bc)
+
+HyperCubic{N, B}(dim::Int) where {N, B <: AbstractBoundary} = HyperCubic{N, B}(dim, B())
+HyperCubic{N}(dim::Int, bc::B=Periodic()) where {N, B <: AbstractBoundary} = HyperCubic{N, B}(dim, bc)
+
+const Chain = HyperCubic{1}
+const Square = HyperCubic{2}
+const Cubic = HyperCubic{3}
+
+_apply_bc(::Type{Periodic}, length, coord) = mod(coord, 1:length)
```
Comment (Roger-luo) [inline src/hypercubic.jl:30]:
```suggestion
_apply_bc(::Periodic, length, coord) = mod(coord, 1:length)
```

Comment (Roger-luo) [reply src/hypercubic.jl:30]:
```julia
apply_boundary(::AbstractBoundary, length::Int, x::Int) = mod(x, 1:length)
apply_boundary(bc::AbstractBoundary, length::Int, x::Coordinate) = Coordinate(apply_boundary.(bc, length, x.coordinates))
```

File: src/boundaries.jl (hunk 7 above)
Comment (emerali) [reply src/boundaries.jl]: gonna skip this since we've decided to store the tuple of BCs in the Lattice struct

**Pull request merged**